### DISCUSSION TITLE
How to prevent hardcoded values on backend from being exposed to client-side bundle?

### DISCUSSION THREAD

**Discussion by tahonaPL** [vercel/next.js#32874] [Help] [answered]
### What version of Next.js are you using?
12.0.6

### What version of Node.js are you using?

16.09

### What browser are you using?

chrome

### What operating system are you using?

windows

### How are you deploying your application?

build and then start

### Describe the Bug

Backend class with constants **is visible in UI** even if variable is not rendered and its purpose is purely backend.

![image](https://user-images.githubusercontent.com/19371479/147559296-21204cd5-969a-4904-a02d-b9b931ebe144.png)
![image](https://user-images.githubusercontent.com/19371479/147559355-eb5aafb0-6b0d-4ebf-aaaf-4e32975cd258.png)
![image](https://user-images.githubusercontent.com/19371479/147559492-a5709c95-8593-4691-b2af-c0a8c40671af.png)

ViewConst.js (Configuration class )contains statics like pageName emaiil etc.   which is used in pages/ to use it contact page etc. 

This is a serious problem that some UI devs doesn't know. Some of them leaked client ids this way.

#### **EDITED**
Situation is even worse all private objects and methods from /pages are leaked:

![image](https://user-images.githubusercontent.com/19371479/148017648-7fc46f87-fe33-4c52-b150-15057b76290d.png)
![image](https://user-images.githubusercontent.com/19371479/148017672-acf56f7b-fd9c-4302-a20a-5f74e46e320e.png)

this behavior affects:
- method functions 
- classes
- objects and statics
- hardcoded queries, selects
- passwords 
- OMG...

I can easily say that this is the **MOST DANGEROUS framework** I ever saw (after WordPress).  

### Expected Behavior

1. Error on build step. 
2. Or some Big Warning that should be disabled by adding file to some exception rules. 
3. Or separate directory for /backend classes.  (+ point 1)

### To Reproduce
written in description

**Comment by balazsorban44** [comment]
Sensitive information should be put into environment variables.

**Reply by tahonaPL** [reply]
Well yes, sensitive info should not leak from backend to UI so easy.
Env variable is common way to pass configuration to app or via KeyVault.  

But not all variables needs to be passed to ENV  some just needs to be hardcoded in backend.  But still don't want them to be added to UI by mistake. Don't want to watch my back constantly. It should be clear decision. not common import like mistake. 

What about functions of backend methods. this will allow to hackers to create exploits.

**Reply by balazsorban44** [reply]
Any env variable not prepended with `NEXT_PUBLIC` will be stripped from the client-side bundle as documented.

If you want to put information somewhere that should not "leak" in any case, just use an environment variable.

**Reply by balazsorban44** [reply]
In any case, I converted your suggestion to an Idea 💡, let's see if it gets traction, we can revisit it then.

**Reply by tahonaPL** [reply]
well it won't get traction. I found my colleagues not aware of the bug and every bigger player don't put things as statics. They as well use some proxy between UI and real backend (Spring/ Express). But this is the worst separation of concerns I ever encounter. I remember that similar issue PHP had when apache files weren't restricted and you could access wordpress config files. let's not allow to have same here.  

Can we ask for some Vercel strong backend architect to look into it?

**Reply by tahonaPL** [reply]
@balazsorban44  Pls move this to bugs (or this https://github.com/vercel/next.js/discussions/25938). As it should be patched. I already found other issues that mention the same problem. To see this as serious problem I think you need to be more backend than UI dev. 

@timneutkens  pls check if this is not a serious problem. I think it is as my UI dev collegues weren't aware of this behavior. (Actually they start to look into it to verify if they don't share too much)

 https://github.com/vercel/next.js/discussions/25938
 https://github.com/vercel/next.js/discussions/32783 
 
 Probably there are more tickets related to this.

**Reply by balazsorban44** [reply]
#32783 seems completely unrelated. please be patient and respect that many are enjoying their holidays.

just to make sure, you say that an unused import is bundled in the client-side code, or that a server-only (read: getStaticProps/getServerSideProps, etc.) import makes it to the client bundle?

Next.js comes with code elimination that would make sure those are separated and that server-side code doesn't make it to the client.

You can check it here https://next-code-elimination.vercel.app/

(I just checked the claim in #25938 and it seems to have been fixed because I don't see the described behavior on the above-mentioned demo site.)


if you have a minimal reproduction of such behavior, indeed it might be an issue.

I would appreciate if you could find out if it's only happening with SWC (the new default compiler) or with Babel as well? (adding a simple `.babelrc` file with `{presets: "next/babel"}` would disable SWC)


I've asked this internally and will get back to you when I have a better answer. 💚

**Reply by tahonaPL** [reply]
Hi @balazsorban44  thx for taking this topic and spending time on that. About  https://github.com/vercel/next.js/discussions/32783  is valid. I have same problem:

"I looked everywhere in the documentation and either I'm blind or can someone explain **how do I write a code that is ONLY accessible to the server.**"
Only is the keyword here. 
And this is my issue here. How I can make sure that one of my teammates or me by mistake not import and use server code in client? (I know - code review.But it is not solution. It already happened that I accepted some stupid stuff on pull request (didn't check the import paths)).

About https://github.com/vercel/next.js/discussions/25938 . well maybe this example with pool is fixed, but If we wrap this into class or have some other method the code is not removed.  Looks like code elimination works only on "getServerSideProps". so the moment I want to extract some part of the code used in "getServerSideProps" to some private method. for example pool.select (...) to => getAllUsers() - this will create security issue and It can be done by mistake in too easy way. 

Taking Murphy law ("Anything that can go wrong will go wrong"). We can assume that devs already share sensitive info to hackers :D in small projects. (sensitive - queries details etc.)

About next/babel could not verify. Some other issues. 

---

I have created some project in anyfiddle:
https://www.anyfiddle.com/p/crownclown67/awwcqfty
https://awwcqfty.anyfiddle.run/  (search in scripts for password)

If you don't want to check the photo:
![image](https://user-images.githubusercontent.com/19371479/147673378-057d2c52-d1ce-45b5-ac7b-ec442e898560.png)
In browser:
![image](https://user-images.githubusercontent.com/19371479/147673658-c6e04f77-9a06-4b8e-8550-b6985b748377.png)

**Reply by balazsorban44** [reply]
For the record, I still think it's bad practice to put sensitive information directly into the code, and that should be saved in environment variables instead. You might get an answer from someone else after the holidays.

**Reply by tahonaPL** [reply]
@balazsorban44 . eh.. if you ask me - env for sensitive data in enterprise solution is a poor choice as well.  But is not about the sensitive data. It is about leaking implementation details and any kind of data to client. For example sql query(potential sql injection), methods validation, method implementations, password, if dev will choose to hard code it (For small one page private project - I would hard code it too for simplicity).

**Reply by pcjmfranken** [reply]
> Sensitive information should be put into environment variables.

This code is never expected or even supposed to be exposed to browsers like this. As @tahonaPL 
 said, implementation details of the functionality that lives here, *is* sensitive information.

The key to your front-door could be secured in a Fort Knox vault, but that matters not if everyone can see you left your back-door wide open.

**Comment by tahonaPL** [comment]
@balazsorban44  any news?

**Reply by tahonaPL** [reply]
@balazsorban44  could you add some strong backend to check this.

**Comment by sicarius97** [comment, marked as answer]
@tahonaPL Some of your implementations above are definitely unsafe, but it isn't because of nextjs. First off, you are exporting functions from an API route so of course they are going to be included in the client side bundle, especially if you are also importing them on a client side page or even if you import it into another API route. If this was a serverless scenario, how else would one lambda share code from another lambda if it doesn't exist before it is instantiated? By accessing it from a shared bundle

Secondly, I completely disagree with the pattern of hard coding any kind of secrets as that is completely unsafe, especially in an enterprise scenario. Environment variables are way way safer, even in an enterprise scenario. Plus you can have a lot of setups where your actual runtime secrets are even hidden from your own devs during development so that they also cant compromise your app What if you hardcode some secrets and one of your devs gos rogue? What if, as in this case, something accidentally ends up client side and you expose your secrets to the public? Only way your env vars can be compromised is in the event someone gains access to your server. At my workplace, we load all of our secrets from a secure secrets manager into the app env or a secrets loader module at buildtime that is ONLY available serverside. If you are trying to implement some modules that need access to server only secrets, you can also use the serverRuntimeConfig from getConfig and everything will only be available serverside.

Lastly to your question: "And this is my issue here. How I can make sure that one of my teammates or me by mistake not import and use server code in client?" The answer is of course code review.. your team can expose secrets or vulnerabilities in all sorts of ways, even in a backend only app. I have seen things leaked through console.logs that weren't taken out, enums and functions that were exported from server routes and could be called from the client (kind of like you did in your above example "vulnerable" implementation), exposing serverside memory to the client, etc etc. Only way all of these things can be prevented is establishing good practices from the start and LOTS AND LOTS of code review

Long story short, I've been using nextjs along with my team for a big internal administration application that has the potential to do a lot of harm if someone were to compromise its security, but I've determined time and time again that it is absolutely secure. I've scrubbed the bundle, used every trick in the book to pull secrets out of memory or try to use internal functions from the client to gain access to backend info, and even worked with an external security agency that determined that the app was completely tight with no known vulnerabilities. I think this really all boils down to how you are writing the app, and not that nextjs is vulnerable itself

**Reply by tahonaPL** [reply]
@sicarius97 Hi thx for checking the issue. Let me answer to your answers:  

1. " First off, you are exporting functions from an API route so of course they are going to be included in the client side bundle"
true, this is just to give the example. But not sure why we need this. No /api code should end up in client code. Is there any reason to have this behavior ? 

About serverless.  "By accessing it from a shared bundle" - well true, but it would be more clear in the project if we consciously decide what library we are adding (not like here adding something by mistake). And this shared library would have client/model/configs name to give clear purpose. different lib would go to UI, different to Backend and shared with no harming and clear purpose code to both (some utils etc.)   

2.  "I completely disagree with the pattern of hard coding any kind of secrets as that is completely unsafe"  - I fully agree with you. But you cannot say that this not happening in the real world. But anyway I wasn't aiming this concerns in my example. It is more about finding holes in the system when you see backend implementation and using this vulnerabilities to retrieve credentials even if they change on hourly basis. 

3.   "The answer is of course code review....your team can expose secrets or vulnerabilities in all sorts of ways" - that is true. But this adds more work for code reviewers. They need to be aware off this problems. 

And let me ask you:
- how often do you check imports ? (ME: rarely, only if I suspect broken concern separation  )
- do you think all teams have high level code review or security company that checks code and all every build?
- what are your other actions to prevent leaking out backend implementation details ? maybe class naming, directory naming (e.g. /backend/..)?  

"Only way all of these things can be prevented is establishing good practices from the start and LOTS AND LOTS of code review" - I don't agree here. Not only way. Of course good practices and code review is very important but I think making something impossible is much better. 

4. "nextjs along with my team for a big internal administration application ...  used every trick in the book to pull secrets out of memory ... worked with an external security agency"  - well I think you are forgetting about smaller players: 1-2 person teams, students, junior projects, small startups, learning projects etc. I don't think they setup pipeline with code vulnerability tests, hire external company to check security or do code reviews on super high quality standard by senior or Tech Lead. 

And yes it is always up to developers how they write apps. But from the other hand it is the role of the framework to be safe and secure as possible.

I rise this topics not because I could not live with this issue it just I don't want to constantly watch my back (and others back too).

**clean solution:**
- adding directory for backend purpose only (configurable and optional  {..backendDir: ["src/"]} ), 
- making sure that backend code cannot be used in other places than /api and backendDir  
- making sure that /api exports cannot leak out (not sure why it is possible. There is no purpose for that.)

**Reply by iiqmark8** [reply]
![image](https://user-images.githubusercontent.com/70702873/148863250-946eacd4-a9d9-4745-8c8a-321ca640470d.jpeg) about pg3d.

**Comment by tahonaPL** [comment]
@balazsorban44 - Unmarking the answer, as the answer tries to justify current implementation and doesn't give real solution. Of course we can leave at that, but It would better if some really strong "backend" from vercel could answer this.

**Reply by leerob** [reply]
@balazsorban44 does understand the backend, and his recommendation to use environment variables is correct. We should not store secrets as plaintext, and especially not commit them to git. 

You mentioned "adding directory for backend purpose only" and "making sure that backend code cannot be used in other places than /api and backendDir". There are already clear conventions in Next.js that denote what code runs on the front end and on the back.

For example, `getStaticProps`, `getStaticPaths`, `getServerSideProps`, API Routes, etc - these run on the server. You can validate the code being removed from the client-side bundle here: https://next-code-elimination.vercel.app/

**Reply by pcjmfranken** [reply]
To embrace a zero trust model is always good advice. Pretty much the perfect non-answer when you really think about it though :stuck_out_tongue_winking_eye:

**Reply by tahonaPL** [reply]
Ehh.... please people. Stop mentioning the env variables, because I start to understand why nextjs is unsafe project.

"balazsorban44 does understand the backend, and his recommendation to use environment variables is correct. We should not store secrets as plaintext, and especially not commit them to git."  - @leerob  - I think he doesn't if this is his only point. 

Anyway  `getServerSideProps` does not solve problems, it give false impression to be safe. The moment we extract code to method or class to separate business and infrastructure code we open our apps for vulnerabilities. Code elimination work only on mentioned method so only solution would be having selects (queries) and business logic in one method, but that would be just stupid and would limit re-usability of the code. 

@pcjmfranken - thx for naming the model. Well we actually doing many things for security that could be attached to "zero trust model". Code reviews, tests, api validation, git master blockade, code security testing, sonar etc. ...

.... And here we need to add one more rule because with nextjs we started to mix things that weren't possible before - we started to mix client code and backend code.
 
@leerob, @pcjmfranken, @balazsorban44  -  Maybe I'm missing some feature that will guard me from using the same code in  backend and client.

**Reply by balazsorban44** [reply]
Yes, you are missing the point of environment variables... There isn't much else from me really, but that should be enough.

@sicarius97 has done a great job summarizing the flaws and explained it in detail how you should solve those issues.

The reason environment variables are desirable in this (and all cases when secure information is needed to be referenced in the source code) case is because they aren't generally bundled with the code. 

In Next.js, only variables prefixed with `NEXT_PUBLIC_` will be. These variables are inlined build time to make them available in the browser bundles.

Environment variables not prefixed with `NEXT_PUBLIC_` aren't inlined in the bundle though, and that's your safest bet, and probably the only truly secure option.

See:
https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser

FYI, Create React App also makes a similar distinction by requiring `REACT_APP_` as a prefix.

https://create-react-app.dev/docs/adding-custom-environment-variables/

Some other examples of similar projects, using the same methodology of prefixing environment variables.

Nuxt.js https://nuxtjs.org/docs/configuration-glossary/configuration-env/#automatic-injection-of-environment-variables

SvelteKit (which uses Vite under the hood)
https://vitejs.dev/guide/env-and-mode.html#env-files

Note, that none of these recommend/mention any kind of secret value to exist in the source code as plain text, for obvious reasons.

Prefixing is safe, because the developer has to be explicit about it, thus avoiding accidental exposure of these values.

You could argue that this is the only zero-trust method of avoiding leaks. You don't even trust Next.js to inline these values, unless you explicitly tell it so by a prefix.

**Reply by leerob** [reply]
Thank you @balazsorban44 - I am locking this discussion as an answer has been provided. Further, there's no need to speak negatively towards him. "I think he doesn't [understand the backend] if this is his only point." is not helpful - please review our [Code of Conduct](https://github.com/vercel/next.js/blob/canary/CODE_OF_CONDUCT.md).
